### PR TITLE
One d root finder

### PR DESCRIPTION
**Pull request by osheamonn**

### CODE REVIEW

File: src/Numerical/RootFinding/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+set(MY_LIBRARY_SOURCES
+    RootFinding/QuadraticEquation.cpp)
```
Comment (fmahebert) [inline src/Numerical/RootFinding/CMakeLists.txt]: Moving the closing paren to the next line may reduce future conflicts (this shamelessly stolen from @nilsdeppe).

File: src/Numerical/RootFinding/QuadraticEquation.cpp [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+#include "ErrorHandling/Assert.hpp"
+#include "ErrorHandling/ExpectsAndEnsures.hpp"
+
+double positive_root(const double a, const double b, const double c) {
+  double x0;
+  double x1;
+  gsl_poly_solve_quadratic(a, b, c, &x0, &x1);
+  Ensures(x0 <= 0.0 and x1 >= 0.0);
+  return x1;
```
Comment (fmahebert) [inline src/Numerical/RootFinding/QuadraticEquation.cpp]: The GSL [documentation](https://www.gnu.org/software/gsl/manual/html_node/Quadratic-Equations.html) indicates that if `gsl_poly_solve_quadratic` finds 0 or 1 real roots, then one or both of `x0`,`x1` will _not_ be modified. If this occurs then `x1` and potentially `x0` remain uninitialized. Spooky.

I would suggest,
* assigning  (`0`, `signaling_nan`, ...?) to both variables at declaration,
* adding an `ASSERT(numrealroots == 2, ...)` as in the function below.

File: src/Numerical/RootFinding/QuadraticEquation.cpp [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+std::array<double, 2> real_roots(const double a, const double b,
+                                 const double c) {
+  double x0;
+  double x1;
+  int numrealroots = gsl_poly_solve_quadratic(a, b, c, &x0, &x1);
+  ASSERT(numrealroots == 2, "Assumes that there are two real roots");
+  return {{x0, x1}};
```
Comment (fmahebert) [inline src/Numerical/RootFinding/QuadraticEquation.cpp]: See above comment re: variable assignment. The `ASSERT` makes this code safe in some builds only.

File: src/Numerical/RootFinding/RootFinder.hpp [hunk 4]
```diff
@@ -0,0 +1,36 @@
+
+/// \file
+/// Declares function find_root_by_brents_method
```
Comment (fmahebert) [inline src/Numerical/RootFinding/RootFinder.hpp]: `find_root_of_function` ?

File: tests/Unit/Numerical/RootFinding/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+set(SPECTRE_UNIT_ROOT_FINDER_TESTS
+  Numerical/RootFinding/Test_OneDRootFinder.cpp
+  Numerical/RootFinding/Test_QuadraticEquation.cpp)
```
Comment (fmahebert) [inline tests/Unit/Numerical/RootFinding/CMakeLists.txt]: Paren

File: tests/Unit/Numerical/RootFinding/Test_OneDRootFinder.cpp [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+namespace {
+double f_free(double x) { return 2.0 - x * x; }
+struct F {
+  double operator()(double x) { return 2.0 - x * x; }
+};
+
+
+TEST_CASE("Unit.Numerical.RootFinding.TOMS748RootSolver", "[Numerical][RootFinding][Unit]") {
+  double abs_tol = 1e-5;
+  double rel_tol = 1e-5;
```
Comment (fmahebert) [inline tests/Unit/Numerical/RootFinding/Test_OneDRootFinder.cpp]: These tolerances seem very loose, I assume to make the test faster? I would suggest adding a comment to explain the reason for these values.

File: src/Numerical/RootFinding/QuadraticEquation.cpp (hunk 2 above)
Comment (nilsdeppe) [reply src/Numerical/RootFinding/QuadraticEquation.cpp]: As long as the GSL doesn't cause the `signaling_nan` to trigger that's the best way to go. If it doesn't play nice (which as a safety precaution we probably don't want to risk) is then we write it as:
```cpp
if (gsl_poly_solve_quadratic(a, b, c, &x0, &x1)) {
  return /*signaling_nan()*/;
}
```
We should make sure that `std::isnan` works as desired with `signaling_nan`

Comment (fmahebert) [reply src/Numerical/RootFinding/QuadraticEquation.cpp]: Maybe you were being loose in your example, but note that the function is unsafe when `gsl_poly_solve_quadratic` returns 0 or 1, and is safe when it returns 2.

File: src/Numerical/RootFinding/QuadraticEquation.hpp [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+/*! \ingroup Functors
+ * \brief Returns the positive root of a quadratic equation ax^2 + bx + c = 0
+ * \returns The positive root of a quadratic equation.
+ * \requires That there are two real roots, of which only one is positive.
+ */
+double positive_root(double a, double b, double c);
+
+/*! \ingroup Functors
+ * \brief Returns the two positive roots of a quadratic equation ax^2 + bx + c =
```
Comment (kidder) [inline src/Numerical/RootFinding/QuadraticEquation.hpp]: two real roots

File: tests/Unit/Numerical/RootFinding/Test_OneDRootFinder.cpp [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  auto root_from_free =
+      find_root_of_function(f_free, lower, upper, abs_tol, rel_tol);
+  auto root_from_functor =
+      find_root_of_function(f_functor, lower, upper, abs_tol, rel_tol);
+  CHECK(std::abs(root_from_lambda - sqrt(2)) < abs_tol);
+  CHECK(std::abs(root_from_lambda - sqrt(2)) / sqrt(2) < rel_tol);
+  CHECK(root_from_free == root_from_lambda);
+  CHECK(root_from_free == root_from_functor);
+}
+}  // namespace
```
Comment (kidder) [inline tests/Unit/Numerical/RootFinding/Test_OneDRootFinder.cpp]: don't put the TEST_CASE in the namespace

File: tests/Unit/Numerical/RootFinding/Test_QuadraticEquation.cpp [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+TEST_CASE("Unit.Numerical.RootFinding.PositiveRoot",
+          "[Numerical][RootFinding][Unit]") {
+  CHECK(Approx(6.31662479035539985) == positive_root(1.0, -6.0, -2.0));
+}
+
+TEST_CASE("Unit.Functors.RealRoots", "[Numerical][RootFinding][Unit]") {
+  auto roots = real_roots(2.0, -11.0, 5.0);
+  CHECK(Approx(0.5) == roots[0]);
+  CHECK(Approx(5.0) == roots[1]);
+}
```
Comment (kidder) [inline tests/Unit/Numerical/RootFinding/Test_QuadraticEquation.cpp]: add a test that has two equal roots
add a test that has one positive, one real root
add a test that has no real roots and triggers the assert

File: tests/Unit/Numerical/RootFinding/Test_QuadraticEquation.cpp [hunk 10]
```diff
@@ -0,0 +1,15 @@
+
+#include <catch.hpp>
+
+#include "Numerical/RootFinding/QuadraticEquation.hpp"
+
+TEST_CASE("Unit.Numerical.RootFinding.PositiveRoot",
+          "[Numerical][RootFinding][Unit]") {
+  CHECK(Approx(6.31662479035539985) == positive_root(1.0, -6.0, -2.0));
+}
+
```
Comment (kidder) [inline tests/Unit/Numerical/RootFinding/Test_QuadraticEquation.cpp]: add a test that has two equal positive roots and triggers the assert
add a test that has two positive roots and triggers the assert
add a test that has no positive roots and triggers the assert

File: src/Numerical/RootFinding/QuadraticEquation.cpp [hunk 11]
```diff
@@ -0,0 +1,26 @@
+
```
Comment (kidder) [inline src/Numerical/RootFinding/QuadraticEquation.cpp]: license

File: src/Numerical/RootFinding/QuadraticEquation.hpp [hunk 12]
```diff
@@ -0,0 +1,22 @@
+
```
Comment (kidder) [inline src/Numerical/RootFinding/QuadraticEquation.hpp]: license

File: src/Numerical/RootFinding/RootFinder.hpp [hunk 13]
```diff
@@ -0,0 +1,36 @@
+
```
Comment (kidder) [inline src/Numerical/RootFinding/RootFinder.hpp]: license

File: tests/Unit/Numerical/RootFinding/Test_OneDRootFinder.cpp [hunk 14]
```diff
@@ -0,0 +1,35 @@
+
```
Comment (kidder) [inline tests/Unit/Numerical/RootFinding/Test_OneDRootFinder.cpp]: license

File: tests/Unit/Numerical/RootFinding/Test_QuadraticEquation.cpp [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
```
Comment (kidder) [inline tests/Unit/Numerical/RootFinding/Test_QuadraticEquation.cpp]: license

File: tests/Unit/Numerical/RootFinding/Test_OneDRootFinder.cpp (hunk 8 above)
Comment (kidder) [inline tests/Unit/Numerical/RootFinding/Test_OneDRootFinder.cpp]: add a test that finds the root at the lower bound and another at the upper bound (this failed in older code)

**Review by kidder — CHANGES_REQUESTED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request closed without merging**